enarx-keep-sgx-shim/src/handler.rs

```diff
@@ -254,31 +253,39 @@ impl<'a> Handler<'a> {
     pub fn read(&mut self) -> u64 {
         self.trace("read", 3);
 
-        let fd = self.aex.gpr.rdi.raw();
-        let buf = self.aex.gpr.rsi.raw() as *mut u8;
-        let size = self.aex.gpr.rdx.raw();
-
-        // Allocate some unencrypted memory.
-        let map = match self.ualloc(size) {
-            Err(errno) => return errno as u64,
-            Ok(map) => map,
+        // Allocate some unencrypted memory from the Block.
+        let cursor = self.block.cursor();
+        let slice: &mut [u8] = match unsafe { cursor.alloc(self.aex.gpr.rdx.into()) } {
+            Ok(slice) => slice,
+            Err(_) => return -libc::EMSGSIZE as u64,
         };
 
-        unsafe {
-            // Do the syscall; replace encrypted memory with unencrypted memory.
-            let ret = self.syscall(libc::SYS_read as u64, fd, map as _, size, 0, 0, 0);
+        // Do the syscall; replace encrypted memory with unencrypted memory.
+        let req = request!(libc::SYS_read => self.aex.gpr.rdi, slice, slice.len());
+        let res = unsafe { self.proxy(req) };
 
-            // Copy the unencrypted input into encrypted memory.
-            if ret <= ERRNO_BASE {
-                if ret > size {
+        match res {
+            Ok(res) => {
+                if usize::from(res[0]) > self.aex.gpr.rdx.into() {
                     self.attacked();
                 }
 
-                copy_nonoverlapping(map, buf, ret as _);
+                let tbuf = unsafe { self.aex.gpr.rsi.as_slice_mut(self.aex.gpr.rdx.into()) };
+
+                // Reallocate some unencrypted memory from the Block.
+                let cursor = self.block.cursor();
+                let slice: &mut [u8] = match unsafe { cursor.alloc(self.aex.gpr.rdx.into()) } {
+                    Ok(slice) => slice,
+                    Err(_) => return -libc::EMSGSIZE as u64,
+                };
```

Comment:
This pattern of:

```rust
let something = match ... {
   Ok(x) => x,
   Err(_e) => return -libc::EMSGSIZE as u64,
}
```

seems to be repeating a lot in this file. Would it make sense to wrap up the implementation of a syscall in a fallible function such as

```rust
fn read(&mut self) -> u64 {
    fn read_impl(&mut self) -> Result<usize, Errno> {
         // the actual impl where you can use '?'
    }

    match self.read_impl() {
        Ok(x) => x,
        Err(e) => u64::from(e),
    }
}
```

Oh, and this might be done in a future PR. I just thought I'd mention it here for posterity.

Reply:
The `Errno` type could be implemented using the `thiserror` crate I guess.

Reply:
@kubkon You're completely right, and this is going away.

Reply:
@kubkon Yes, that is precisely where we are going. See my latest PR which implements this.